In [1]:
from os import listdir
from os.path import isfile, join
mypath = './dataset'
onlyfiles = [f for f in listdir(mypath) if isfile(join(mypath, f))]
onlyfiles

['batt_info.csv000',
 'devuse_4known_device.csv000',
 'mousewait.csv000.gz',
 'system_sysinfo_unique_normalized.csv000.gz',
 'plist_process_resource_util_hist.csv003.gz',
 'plist_process_resource_util_hist.csv002.gz',
 'plist_process_resource_util_hist.csv001.gz',
 'batt_acdc_events.csv000',
 'devuse_4known_device.csv001',
 'frgnd_v2.csv000',
 'frgnd_backgrnd_apps.csv000',
 'plist_process_resource_util_13wks.csv000.gz',
 'hw_metric_histo.csv001.gz',
 'hw_metric_histo.csv000.gz',
 'plist_process_resource_util_13wks.csv001.gz']

In [2]:
'frgnd_backgrnd_apps.csv000.gz',
'devuse_4known_device.csv000.gz',
'devuse_4known_device.csv001.gz',
'batt_acdc_events.csv000.gz',
'batt_info.csv000.gz',

('batt_info.csv000.gz',)

# EDA

## Loading data into DataFrames

In [3]:
import dask.dataframe as dd
from dask.diagnostics import ProgressBar 
from matplotlib import pyplot as plt
from dask.delayed import delayed

import pandas as pd
import os


list_of_fp=['frgnd_backgrnd_apps.csv000.gz',
'devuse_4known_device.csv000.gz',
'devuse_4known_device.csv001.gz',
'batt_acdc_events.csv000.gz',
'batt_info.csv000.gz']


#dfs = [delayed(pd.read_csv)(mypath+"/"+fn) for fn in list_of_fp]

#df = dd.from_delayed(dfs)

fp=mypath+"/"+'frgnd_backgrnd_apps.csv000'
fp2 = mypath+"/"+'devuse_4known_device.csv000'
fp3 = mypath+"/"+'devuse_4known_device.csv001'
fp4 = mypath+"/"+'batt_acdc_events.csv000'
fp5 = mypath+"/"+'batt_info.csv000'

frgnd_backgrnd_apps = dd.read_csv(fp,sep='')
devuse1 = dd.read_csv(fp2,sep='\t',error_bad_lines=False)
devuse2 = dd.read_csv(fp3,sep='\t',error_bad_lines=False,dtype={'duration': 'float64'})
batt_acdc = dd.read_csv(fp4,sep='\1').head(500000)
batt_info = dd.read_csv(fp5,sep='\1',dtype={'designed_capacity': 'float64',
       'full_charge_capacity': 'float64'})



/opt/conda/lib/python3.7/site-packages/dask/dataframe/core.py:6049: UserWarning: Insufficient elements for `head`. 500000 elements requested, only 206369 elements available. Try passing larger `npartitions` to `head`.
  warnings.warn(msg.format(n, len(r)))


In [4]:
batt_info

Dask DataFrame Structure:
               load_ts batch_id audit_zip audit_internal_path    guid interval_start_utc interval_end_utc interval_local_start interval_local_end      ts      dt battery_enum chemistry designed_capacity full_charge_capacity battery_count
npartitions=86                                                                                                                                                                                                                               
                object   object    object              object  object             object           object               object             object  object  object        int64    object           float64              float64         int64
                   ...      ...       ...                 ...     ...                ...              ...                  ...                ...     ...     ...          ...       ...               ...                  ...           ...
...                ...      ...       ...                 ...     ...                ...              ...                  ...                ...     ...     ...          ...       ...               ...                  ...           ...
                   ...      ...       ...                 ...     ...                ...              ...                  ...                ...     ...     ...          ...       ...               ...                  ...           ...
                   ...      ...       ...                 ...     ...                ...              ...                  ...                ...     ...     ...          ...       ...               ...                  ...           ...
Dask Name: from-delayed, 258 tasks

In [125]:
batt_acdc

load_ts         batch_id  \
0       2020-04-23 04:07:48  20200422-210232   
1       2020-04-23 04:07:48  20200422-210232   
2       2020-04-23 04:07:48  20200422-210232   
3       2020-04-23 04:07:48  20200422-210232   
4       2020-04-23 04:07:48  20200422-210232   
...                     ...              ...   
206364  2020-05-11 04:03:59  20200510-200229   
206365  2020-05-11 04:03:59  20200510-200229   
206366  2020-05-11 04:03:59  20200510-200229   
206367  2020-05-11 04:03:59  20200510-200229   
206368  2020-05-11 04:03:59  20200510-200229   

                                                audit_zip  \
0       2020042207-i-04a9a906e1cd5620c-kkPOjcyZXZqMOox...   
1       2020042207-i-04a9a906e1cd5620c-kkPOjcyZXZqMOox...   
2       2020042207-i-04a9a906e1cd5620c-kkPOjcyZXZqMOox...   
3       2020042207-i-04a9a906e1cd5620c-kkPOjcyZXZqMOox...   
4       2020042207-i-04a9a906e1cd5620c-kkPOjcyZXZqMOox...   
...                                                   ...   
206364  2020051016-i-0728cf8ab79db9e58-OwS0ZQVNaswIZq5...   
206365  2020051016-i-0728cf8ab79db9e58-OwS0ZQVNaswIZq5...   
206366  2020051016-i-0728cf8ab79db9e58-OwS0ZQVNaswIZq5...   
206367  2020051016-i-0728cf8ab79db9e58-OwS0ZQVNaswIZq5...   
206368  2020051016-i-0728cf8ab79db9e58-OwS0ZQVNaswIZq5...   

                   audit_internal_path                              guid  \
0       V8BATTEVENTS_20200422093417.V8  000037b573c4463abc8cba503ac06045   
1       V8BATTEVENTS_20200422093417.V8  000037b573c4463abc8cba503ac06045   
2       V8BATTEVENTS_20200422093417.V8  000037b573c4463abc8cba503ac06045   
3       V8BATTEVENTS_20200422093417.V8  000037b573c4463abc8cba503ac06045   
4       V8BATTEVENTS_20200422093417.V8  000037b573c4463abc8cba503ac06045   
...                                ...                               ...   
206364  V8BATTEVENTS_20200510105945.V8  0001a9100b1846b0a8ebf15ce1c9bf4c   
206365  V8BATTEVENTS_20200510105945.V8  0001a9100b1846b0a8ebf15ce1c9bf4c   
206366  V8BATTEVENTS_20200510105945.V8  0001a9100b1846b0a8ebf15ce1c9bf4c   
206367  V8BATTEVENTS_20200510105945.V8  0001a9100b1846b0a8ebf15ce1c9bf4c   
206368  V8BATTEVENTS_20200510105945.V8  0001a9100b1846b0a8ebf15ce1c9bf4c   

         interval_start_utc     interval_end_utc interval_local_start  \
0       2020-04-21 07:32:56  2020-04-22 07:34:41  2020-04-21 09:32:56   
1       2020-04-21 07:32:56  2020-04-22 07:34:41  2020-04-21 09:32:56   
2       2020-04-21 07:32:56  2020-04-22 07:34:41  2020-04-21 09:32:56   
3       2020-04-21 07:32:56  2020-04-22 07:34:41  2020-04-21 09:32:56   
4       2020-04-21 07:32:56  2020-04-22 07:34:41  2020-04-21 09:32:56   
...                     ...                  ...                  ...   
206364  2020-05-09 13:04:37  2020-05-10 15:59:55  2020-05-09 08:04:37   
206365  2020-05-09 13:04:37  2020-05-10 15:59:55  2020-05-09 08:04:37   
206366  2020-05-09 13:04:37  2020-05-10 15:59:55  2020-05-09 08:04:37   
206367  2020-05-09 13:04:37  2020-05-10 15:59:55  2020-05-09 08:04:37   
206368  2020-05-09 13:04:37  2020-05-10 15:59:55  2020-05-09 08:04:37   

         interval_local_end                   ts          dt  \
0       2020-04-22 09:34:41  2020-04-21 09:33:26  2020-04-21   
1       2020-04-22 09:34:41  2020-04-21 09:46:56  2020-04-21   
2       2020-04-22 09:34:41  2020-04-21 09:47:10  2020-04-21   
3       2020-04-22 09:34:41  2020-04-21 09:50:03  2020-04-21   
4       2020-04-22 09:34:41  2020-04-21 20:31:15  2020-04-21   
...                     ...                  ...         ...   
206364  2020-05-10 10:59:55  2020-05-09 11:12:20  2020-05-09   
206365  2020-05-10 10:59:55  2020-05-09 11:28:19  2020-05-09   
206366  2020-05-10 10:59:55  2020-05-09 14:19:43  2020-05-09   
206367  2020-05-10 10:59:55  2020-05-09 15:35:33  2020-05-09   
206368  2020-05-10 10:59:55  2020-05-09 16:10:51  2020-05-09   

       system_power_state event_type   duration  battery_percent_remaining  \
0                 POWERON         DC    810.408                         7

In [6]:
#Inspecting the columns of a DataFrame
print(frgnd_backgrnd_apps.columns)
print(devuse1.columns)
print(devuse2.columns)
print(batt_acdc.columns)
print(batt_info.columns)

Index(['load_ts', 'batch_id', 'audit_zip', 'audit_internal_path', 'guid',
       'interval_start_utc', 'interval_end_utc', 'interval_local_start',
       'interval_local_end', 'event_dt', 'event_ts', 'event_name',
       'event_duration_ms', 'num_proc_samples', 'frgnd_proc_name',
       'frgnd_proc_dt', 'frgnd_proc_ts', 'frgnd_proc_duration_ms',
       'frgnd_proc_pkg', 'frgnd_captioned', 'frgnd_window_mode',
       'frgnd_non_responsive', 'proc_name', 'proc_cpu_utilization'],
      dtype='object')
Index(['load_ts', 'batch_id', 'audit_zip', 'audit_internal_path', 'guid',
       'interval_start_utc', 'interval_end_utc', 'interval_local_start',
       'interval_local_end', 'ts', 'dt', 'device', 'hw_name', 'name',
       'duration', 'status'],
      dtype='object')
Index(['load_ts', 'batch_id', 'audit_zip', 'audit_internal_path', 'guid',
       'interval_start_utc', 'interval_end_utc', 'interval_local_start',
       'interval_local_end', 'ts', 'dt', 'device', 'hw_name', 'name',
       'du

In [7]:
#Finding the common columns between the  DataFrames
from functools import reduce

columns = [set(frgnd_backgrnd_apps.columns),
    set(devuse1.columns),
    set(devuse2.columns),
    set(batt_acdc.columns),
          set(batt_info.columns)]
common_columns = list(reduce(lambda a, i: a.intersection(i), columns))
common_columns 

['guid',
 'interval_local_end',
 'load_ts',
 'interval_start_utc',
 'interval_local_start',
 'audit_internal_path',
 'batch_id',
 'audit_zip',
 'interval_end_utc']

Now that we have a set of common columns shared by all four of the DataFrames, let’s take a look at the first couple of rows of the frgnd_backgrnd_apps DataFrame.

In [8]:
frgnd_backgrnd_apps.head()

load_ts         batch_id  \
0  2020-10-12 03:07:56  20201011-200228   
1  2020-10-12 03:07:56  20201011-200228   
2  2020-10-12 03:07:56  20201011-200228   
3  2020-10-12 03:07:56  20201011-200228   
4  2020-10-12 03:07:56  20201011-200228   

                                           audit_zip  \
0  2020101115-i-040171543e1bfef37-ud0Qz3DtGaeMoEs...   
1  2020101115-i-040171543e1bfef37-ud0Qz3DtGaeMoEs...   
2  2020101115-i-040171543e1bfef37-ud0Qz3DtGaeMoEs...   
3  2020101115-i-040171543e1bfef37-ud0Qz3DtGaeMoEs...   
4  2020101115-i-040171543e1bfef37-ud0Qz3DtGaeMoEs...   

                         audit_internal_path  \
0  V8_1_FRGND_BACKGRND_APPS20201010201454.V8   
1  V8_1_FRGND_BACKGRND_APPS20201010201454.V8   
2  V8_1_FRGND_BACKGRND_APPS20201010201454.V8   
3  V8_1_FRGND_BACKGRND_APPS20201010201454.V8   
4  V8_1_FRGND_BACKGRND_APPS20201010201454.V8   

                               guid   interval_start_utc     interval_end_utc  \
0  003019b8c0ae454eb228609aa48ac8d1  2020-10-10 00:14:10  2020-10-11 00:15:17   
1  003019b8c0ae454eb228609aa48ac8d1  2020-10-10 00:14:10  2020-10-11 00:15:17   
2  003019b8c0ae454eb228609aa48ac8d1  2020-10-10 00:14:10  2020-10-11 00:15:17   
3  003019b8c0ae454eb228609aa48ac8d1  2020-10-10 00:14:10  2020-10-11 00:15:17   
4  003019b8c0ae454eb228609aa48ac8d1  2020-10-10 00:14:10  2020-10-11 00:15:17   

  interval_local_start   interval_local_end    event_dt  ... frgnd_proc_name  \
0  2020-10-09 20:14:10  2020-10-10 20:15:17  2020-10-09  ...             NaN   
1  2020-10-09 20:14:10  2020-10-10 20:15:17  2020-10-09  ...      chrome.exe   
2  2020-10-09 20:14:10  2020-10-10 20:15:17  2020-10-09  ...      chrome.exe   
3  2020-10-09 20:14:10  2020-10-10 20:15:17  2020-10-09  ...             NaN   
4  2020-10-09 20:14:10  2020-10-10 20:15:17  2020-10-09  ...      chrome.exe   

  frgnd_proc_dt            frgnd_proc_ts  frgnd_proc_duration_ms  \
0    2020-10-09  2020-10-09 22:00:25.561                       0   
1    2020-10-09  2020-10-09 22:28:43.737                  802973   
2    2020-10-09   2020-10-09 22:50:41.24                 2343497   
3    2020-10-09  2020-10-09 20:14:48.094                       0   
4    2020-10-09   2020-10-09 22:50:41.24                 2343497   

  frgnd_proc_pkg frgnd_captioned frgnd_window_mode  frgnd_non_responsive  \
0            NaN             NaN               NaN                     f   
1            NaN              c1                 W                     f   
2            NaN              c1                 W                     f   
3            NaN             NaN               NaN                     f   
4            NaN              c1                 W                     f   

                proc_name proc_cpu_utilization  
0             DSATray.exe                 0.12  
1  GameScannerService.exe                 0.39  
2  GameScannerService.exe                 0.85  
3     MoUsoCoreWorker.exe                 1.12  
4             MsMpEng.exe                 0.58  

[5 rows x 24 columns]

In [9]:
#Looking at the head of the frgnd_backgrnd_apps DataFrame
frgnd_backgrnd_apps[common_columns].head()

guid   interval_local_end              load_ts  \
0  003019b8c0ae454eb228609aa48ac8d1  2020-10-10 20:15:17  2020-10-12 03:07:56   
1  003019b8c0ae454eb228609aa48ac8d1  2020-10-10 20:15:17  2020-10-12 03:07:56   
2  003019b8c0ae454eb228609aa48ac8d1  2020-10-10 20:15:17  2020-10-12 03:07:56   
3  003019b8c0ae454eb228609aa48ac8d1  2020-10-10 20:15:17  2020-10-12 03:07:56   
4  003019b8c0ae454eb228609aa48ac8d1  2020-10-10 20:15:17  2020-10-12 03:07:56   

    interval_start_utc interval_local_start  \
0  2020-10-10 00:14:10  2020-10-09 20:14:10   
1  2020-10-10 00:14:10  2020-10-09 20:14:10   
2  2020-10-10 00:14:10  2020-10-09 20:14:10   
3  2020-10-10 00:14:10  2020-10-09 20:14:10   
4  2020-10-10 00:14:10  2020-10-09 20:14:10   

                         audit_internal_path         batch_id  \
0  V8_1_FRGND_BACKGRND_APPS20201010201454.V8  20201011-200228   
1  V8_1_FRGND_BACKGRND_APPS20201010201454.V8  20201011-200228   
2  V8_1_FRGND_BACKGRND_APPS20201010201454.V8  20201011-200228   
3  V8_1_FRGND_BACKGRND_APPS20201010201454.V8  20201011-200228   
4  V8_1_FRGND_BACKGRND_APPS20201010201454.V8  20201011-200228   

                                           audit_zip     interval_end_utc  
0  2020101115-i-040171543e1bfef37-ud0Qz3DtGaeMoEs...  2020-10-11 00:15:17  
1  2020101115-i-040171543e1bfef37-ud0Qz3DtGaeMoEs...  2020-10-11 00:15:17  
2  2020101115-i-040171543e1bfef37-ud0Qz3DtGaeMoEs...  2020-10-11 00:15:17  
3  2020101115-i-040171543e1bfef37-ud0Qz3DtGaeMoEs...  2020-10-11 00:15:17  
4  2020101115-i-040171543e1bfef37-ud0Qz3DtGaeMoEs...  2020-10-11 00:15:17

## Cleaning and transforming DataFrames

In [10]:
frgnd_backgrnd_apps_poi = frgnd_backgrnd_apps.columns.difference(common_columns)
frgnd_backgrnd_apps_poi

Index(['event_dt', 'event_duration_ms', 'event_name', 'event_ts',
       'frgnd_captioned', 'frgnd_non_responsive', 'frgnd_proc_dt',
       'frgnd_proc_duration_ms', 'frgnd_proc_name', 'frgnd_proc_pkg',
       'frgnd_proc_ts', 'frgnd_window_mode', 'num_proc_samples',
       'proc_cpu_utilization', 'proc_name'],
      dtype='object')

In [11]:
frgnd_backgrnd_apps[frgnd_backgrnd_apps_poi].head()

event_dt  event_duration_ms      event_name                 event_ts  \
0  2020-10-09             156317  DC_DISPLAY_OFF  2020-10-09 22:00:25.561   
1  2020-10-09            5203164   DC_DISPLAY_ON  2020-10-09 22:03:01.879   
2  2020-10-09            5203164   DC_DISPLAY_ON  2020-10-09 22:03:01.879   
3  2020-10-09            4773068   DC_DISPLAY_ON  2020-10-09 20:14:48.094   
4  2020-10-09            5203164   DC_DISPLAY_ON  2020-10-09 22:03:01.879   

  frgnd_captioned frgnd_non_responsive frgnd_proc_dt  frgnd_proc_duration_ms  \
0             NaN                    f    2020-10-09                       0   
1              c1                    f    2020-10-09                  802973   
2              c1                    f    2020-10-09                 2343497   
3             NaN                    f    2020-10-09                       0   
4              c1                    f    2020-10-09                 2343497   

  frgnd_proc_name frgnd_proc_pkg            frgnd_proc_ts frgnd_window_mode  \
0             NaN            NaN  2020-10-09 22:00:25.561               NaN   
1      chrome.exe            NaN  2020-10-09 22:28:43.737                 W   
2      chrome.exe            NaN   2020-10-09 22:50:41.24                 W   
3             NaN            NaN  2020-10-09 20:14:48.094               NaN   
4      chrome.exe            NaN   2020-10-09 22:50:41.24                 W   

   num_proc_samples  proc_cpu_utilization               proc_name  
0                16                  0.12             DSATray.exe  
1                80                  0.39  GameScannerService.exe  
2               234                  0.85  GameScannerService.exe  
3               478                  1.12     MoUsoCoreWorker.exe  
4               234                  0.58             MsMpEng.exe

In [12]:
devuse1_poi = devuse1.columns.difference(common_columns)
devuse1_poi

Index(['device', 'dt', 'duration', 'hw_name', 'name', 'status', 'ts'], dtype='object')

In [13]:
devuse1[devuse1_poi].head()

device          dt  duration  \
0  GUID_DEVINTERFACE_DISPLAY_ADAPTER  2019-11-10     16571   
1         GUID_DEVINTERFACE_KEYBOARD  2019-11-11     82117   
2            GUID_DEVINTERFACE_MOUSE  2019-11-11     82117   
3                GUID_DEVICE_BATTERY  2019-11-11     82117   
4              GUID_DEVINTERFACE_HID  2019-11-11     82117   

                                        hw_name  \
0  PCI\VEN_8086&DEV_0F31&SUBSYS_14DD1043&REV_0E   
1                         ACPI\VEN_ATK&DEV_3001   
2                         ACPI\VEN_ETD&DEV_0108   
3                         ACPI\VEN_PNP&DEV_0C0A   
4                          HID\VEN_ATK&DEV_4001   

                                              name  status  \
0                             Intel(R) HD Graphics       4   
1       PC/AT Enhanced PS/2 Keyboard (101/102-Key)       4   
2                                    ASUS Touchpad       4   
3  Microsoft ACPI-Compliant Control Method Battery       4   
4            HID-compliant wireless radio controls       4   

                    ts  
0  2019-11-10 20:58:51  
1  2019-11-11 21:00:33  
2  2019-11-11 21:00:33  
3  2019-11-11 21:00:33  
4  2019-11-11 21:00:33

In [14]:
devuse2_poi = devuse2.columns.difference(common_columns)
devuse2[devuse2_poi].head()

device          dt  duration  \
0        GUID_DEVINTERFACE_DISK  2020-05-10   24953.0   
1      GUID_DEVINTERFACE_VOLUME  2020-05-10   24953.0   
2       GUID_DEVINTERFACE_IMAGE  2020-05-10   24953.0   
3      GUID_DEVINTERFACE_SENSOR  2020-05-10   24953.0   
4  GUID_DEVINTERFACE_USB_DEVICE  2020-05-10   24953.0   

                                 hw_name  \
0  SCSI\DiskSAMSUNG_MZNLN512HMJP-000MAV2   
1                         STORAGE\Volume   
2         SUSD_VISTA_2008_WIN7_2008R2_64   
3         Sensors\SensorsSimulatorDriver   
4         USB\VID_04F2&PID_B52C&REV_0029   

                                                name  status  \
0                         SAMSUNG MZNLN512HMJP-000L7       4   
1                                             Volume       4   
2                      Samsung Universal Scan Driver       4   
3  Microsoft Visual Studio Location Simulator Sensor       4   
4                               USB Composite Device       4   

                    ts  
0  2020-05-10 22:48:53  
1  2020-05-10 22:48:53  
2  2020-05-10 22:48:53  
3  2020-05-10 22:48:53  
4  2020-05-10 22:48:53

In [15]:
batt_acdc_poi = batt_acdc.columns.difference(common_columns)
batt_acdc[batt_acdc_poi].head()

battery_minutes_remaining  battery_percent_remaining          dt  \
0                  31.616667                         75  2020-04-21   
1                  73.766670                         69  2020-04-21   
2                  73.766670                         69  2020-04-21   
3                 503.700012                         67  2020-04-21   
4                   0.000000                          8  2020-04-21   

    duration event_type system_power_state                   ts  
0    810.408         DC            POWERON  2020-04-21 09:33:26  
1     13.573         DC           POWEROFF  2020-04-21 09:46:56  
2    173.669         DC            POWERON  2020-04-21 09:47:10  
3  38471.500         DC            POWERON  2020-04-21 09:50:03  
4  43907.823         DC           POWEROFF  2020-04-21 20:31:15

In [16]:
batt_info_poi = batt_info.columns.difference(common_columns)
batt_info[batt_info_poi].head()

battery_count  battery_enum chemistry  designed_capacity          dt  \
0              1             1      LION            43092.0  2020-03-02   
1              1             1      LION            43092.0  2020-03-02   
2              1            -1       NaN            43092.0  2020-03-03   
3              1             1      LION            43092.0  2020-03-03   
4              1            -1       NaN            43092.0  2020-03-03   

   full_charge_capacity                   ts  
0               43320.0  2020-03-02 12:54:05  
1               43320.0  2020-03-02 12:57:12  
2               43320.0  2020-03-03 08:09:34  
3               43320.0  2020-03-03 13:06:26  
4               43320.0  2020-03-03 16:46:17

In order to help Dask read our dataset correctly, we’ll need to manually define a schema for our data instead of relying on type inference.

### Dealing with missing values

In [22]:
#Calculating the percentage of missing values by column
missing_values_frgnd_backgrnd = frgnd_backgrnd_apps.isnull().sum()

with ProgressBar():
    percent_missing_frgnd_backgrnd = ((missing_values_frgnd_backgrnd / frgnd_backgrnd_apps.index.size) * 100).compute()
percent_missing_frgnd_backgrnd

[########################################] | 100% Completed | 28.5s


load_ts                    0.000000
batch_id                   0.000000
audit_zip                  0.000000
audit_internal_path        0.000000
guid                       0.000000
interval_start_utc         0.000000
interval_end_utc           0.000000
interval_local_start       0.000000
interval_local_end         0.000000
event_dt                   0.000000
event_ts                   0.000000
event_name                 0.000000
event_duration_ms          0.000000
num_proc_samples           0.000000
frgnd_proc_name           11.777160
frgnd_proc_dt              0.000000
frgnd_proc_ts              0.000000
frgnd_proc_duration_ms     0.000000
frgnd_proc_pkg            98.686878
frgnd_captioned           12.805952
frgnd_window_mode         12.805952
frgnd_non_responsive       0.000000
proc_name                  0.000000
proc_cpu_utilization       0.000000
dtype: float64

In [23]:
#missing_values_devuse1 = devuse1.isnull().sum()

#with ProgressBar():
    #percent_missing_devuse1 = ((missing_values_devuse1 / devuse1.index.size) * 100).compute()
#percent_missing_devuse1

In [24]:
#missing_values_devuse2 = devuse2.isnull().sum()

#with ProgressBar():
    #percent_missing_devuse2 = ((missing_values_devuse2 / devuse2.index.size) * 100).compute()
#percent_missing_devuse2

In [28]:
missing_values_batt_acdc = batt_acdc.isnull().sum()
missing_values_batt_acdc

load_ts                      0
batch_id                     0
audit_zip                    0
audit_internal_path          0
guid                         0
interval_start_utc           0
interval_end_utc             0
interval_local_start         0
interval_local_end           0
ts                           0
dt                           0
system_power_state           0
event_type                   0
duration                     0
battery_percent_remaining    0
battery_minutes_remaining    0
dtype: int64

In [31]:
missing_values_batt_info = batt_info.isnull().sum()

with ProgressBar():
    percent_missing_batt_info = ((missing_values_batt_info / batt_info.index.size) * 100).compute()
percent_missing_batt_info

[########################################] | 100% Completed |  1min 57.4s


load_ts                  0.000000
batch_id                 0.000000
audit_zip                0.000000
audit_internal_path      0.000000
guid                     0.000000
interval_start_utc       0.000000
interval_end_utc         0.000000
interval_local_start     0.000000
interval_local_end       0.000000
ts                       0.000000
dt                       0.000000
battery_enum             0.000000
chemistry               85.777487
designed_capacity        0.000000
full_charge_capacity     0.000000
battery_count            0.000000
dtype: float64

###  DROPPING COLUMNS WITH MISSING VALUES

Now that we know what we have to work with, we’ll start by dropping any columns that most of their values missing.
We don’t likely stand to gain much information by keeping such a sparse column around, so we’ll remove it from our dataset.

In [35]:
#Dropping columns that have more than 90% missing values
columns_to_drop_frgnd_backgrnd = list(percent_missing_frgnd_backgrnd[percent_missing_frgnd_backgrnd >= 90].index)
frgnd_backgrnd_clean_stage2 = frgnd_backgrnd_apps.drop(columns_to_drop_frgnd_backgrnd, axis=1)

In [36]:
columns_to_drop_frgnd_backgrnd

['frgnd_proc_pkg']

In [37]:
#Dropping columns that have more than 90% missing values
#columns_to_drop_devuse1 = list(percent_missing_devuse1[percent_missing_devuse1 >= 90].index)
#devuse1_clean_stage1 = devuse1.drop(columns_to_drop_devuse1, axis=1)

In [38]:
#columns_to_drop_devuse1

In [39]:
#Dropping columns that have more than 90% missing values
#columns_to_drop_devuse2 = list(percent_missing_devuse2[percent_missing_devuse2 >= 90].index)
#devuse2_clean_stage1 = devuse2.drop(columns_to_drop_devuse2, axis=1)

In [40]:
#columns_to_drop_devuse2

In [41]:
#Dropping columns that have more than 90% missing values
#columns_to_drop_batt_acdc = list(percent_missing_batt_acdc[percent_missing_batt_acdc>= 90].index)
#batt_acdc_clean_stage1 = batt_acdc.drop(columns_to_drop_batt_acdc, axis=1)

In [42]:
#columns_to_drop_batt_acdc

In [43]:
#Dropping columns that have more than 90% missing values
#columns_to_drop_batt_info = list(percent_missing_batt_info[percent_missing_batt_info >= 90].index)
##batt_info_clean_stage1 = batt_info.drop(columns_to_drop_batt_info, axis=1)

In [44]:
#columns_to_drop_batt_info

### Parse Date

In [45]:
frgnd_backgrnd_apps_poi=frgnd_backgrnd_clean_stage2.columns.difference(common_columns)
frgnd_backgrnd_apps_poi

Index(['event_dt', 'event_duration_ms', 'event_name', 'event_ts',
       'frgnd_captioned', 'frgnd_non_responsive', 'frgnd_proc_dt',
       'frgnd_proc_duration_ms', 'frgnd_proc_name', 'frgnd_proc_ts',
       'frgnd_window_mode', 'num_proc_samples', 'proc_cpu_utilization',
       'proc_name'],
      dtype='object')

In [46]:
frgnd_backgrnd_clean_stage2[frgnd_backgrnd_apps_poi].head()

event_dt  event_duration_ms      event_name                 event_ts  \
0  2020-10-09             156317  DC_DISPLAY_OFF  2020-10-09 22:00:25.561   
1  2020-10-09            5203164   DC_DISPLAY_ON  2020-10-09 22:03:01.879   
2  2020-10-09            5203164   DC_DISPLAY_ON  2020-10-09 22:03:01.879   
3  2020-10-09            4773068   DC_DISPLAY_ON  2020-10-09 20:14:48.094   
4  2020-10-09            5203164   DC_DISPLAY_ON  2020-10-09 22:03:01.879   

  frgnd_captioned frgnd_non_responsive frgnd_proc_dt  frgnd_proc_duration_ms  \
0             NaN                    f    2020-10-09                       0   
1              c1                    f    2020-10-09                  802973   
2              c1                    f    2020-10-09                 2343497   
3             NaN                    f    2020-10-09                       0   
4              c1                    f    2020-10-09                 2343497   

  frgnd_proc_name            frgnd_proc_ts frgnd_window_mode  \
0             NaN  2020-10-09 22:00:25.561               NaN   
1      chrome.exe  2020-10-09 22:28:43.737                 W   
2      chrome.exe   2020-10-09 22:50:41.24                 W   
3             NaN  2020-10-09 20:14:48.094               NaN   
4      chrome.exe   2020-10-09 22:50:41.24                 W   

   num_proc_samples  proc_cpu_utilization               proc_name  
0                16                  0.12             DSATray.exe  
1                80                  0.39  GameScannerService.exe  
2               234                  0.85  GameScannerService.exe  
3               478                  1.12     MoUsoCoreWorker.exe  
4               234                  0.58             MsMpEng.exe

In [47]:
with ProgressBar():
    frgnd_backgrnd_event_dt_value_count = frgnd_backgrnd_clean_stage2['event_dt'].value_counts().compute()

[########################################] | 100% Completed | 23.8s


In [65]:
import seaborn as sns
frgnd_backgrnd_event_dt_value_count

2020-10-22    70024
2020-10-14    69961
2020-10-15    69882
2020-10-21    69882
2020-10-19    69163
              ...  
2021-01-02    12338
2021-01-01    11075
2021-01-07     6497
2020-10-09      881
2021-01-08       10
Name: event_dt, Length: 92, dtype: int64

#### Finding1: Most of the data is in 10-2020

In [93]:
frgnd_backgrnd_clean_stage2['event_dt']=dd.to_datetime(frgnd_backgrnd_clean_stage2.event_dt,unit='ns')
frgnd_backgrnd_clean_stage2["frgnd_proc_dt"]=dd.to_datetime(frgnd_backgrnd_clean_stage2.frgnd_proc_dt,unit='ns')
frgnd_backgrnd_clean_stage2['event_ts']=dd.to_datetime(frgnd_backgrnd_clean_stage2.event_ts,unit='ns')
frgnd_backgrnd_clean_stage2["frgnd_proc_ts"]=dd.to_datetime(frgnd_backgrnd_clean_stage2.frgnd_proc_ts,unit='ns')

In [92]:
frgnd_backgrnd_clean_stage2.head()["event_ts"]

0   2020-10-09 22:00:25.561
1   2020-10-09 22:03:01.879
2   2020-10-09 22:03:01.879
3   2020-10-09 20:14:48.094
4   2020-10-09 22:03:01.879
Name: event_ts, dtype: datetime64[ns]

## Summary statistic 

In [96]:
frgnd_backgrnd_clean_stage2.head()

load_ts         batch_id  \
0  2020-10-12 03:07:56  20201011-200228   
1  2020-10-12 03:07:56  20201011-200228   
2  2020-10-12 03:07:56  20201011-200228   
3  2020-10-12 03:07:56  20201011-200228   
4  2020-10-12 03:07:56  20201011-200228   

                                           audit_zip  \
0  2020101115-i-040171543e1bfef37-ud0Qz3DtGaeMoEs...   
1  2020101115-i-040171543e1bfef37-ud0Qz3DtGaeMoEs...   
2  2020101115-i-040171543e1bfef37-ud0Qz3DtGaeMoEs...   
3  2020101115-i-040171543e1bfef37-ud0Qz3DtGaeMoEs...   
4  2020101115-i-040171543e1bfef37-ud0Qz3DtGaeMoEs...   

                         audit_internal_path  \
0  V8_1_FRGND_BACKGRND_APPS20201010201454.V8   
1  V8_1_FRGND_BACKGRND_APPS20201010201454.V8   
2  V8_1_FRGND_BACKGRND_APPS20201010201454.V8   
3  V8_1_FRGND_BACKGRND_APPS20201010201454.V8   
4  V8_1_FRGND_BACKGRND_APPS20201010201454.V8   

                               guid   interval_start_utc     interval_end_utc  \
0  003019b8c0ae454eb228609aa48ac8d1  2020-10-10 00:14:10  2020-10-11 00:15:17   
1  003019b8c0ae454eb228609aa48ac8d1  2020-10-10 00:14:10  2020-10-11 00:15:17   
2  003019b8c0ae454eb228609aa48ac8d1  2020-10-10 00:14:10  2020-10-11 00:15:17   
3  003019b8c0ae454eb228609aa48ac8d1  2020-10-10 00:14:10  2020-10-11 00:15:17   
4  003019b8c0ae454eb228609aa48ac8d1  2020-10-10 00:14:10  2020-10-11 00:15:17   

  interval_local_start   interval_local_end   event_dt  ... num_proc_samples  \
0  2020-10-09 20:14:10  2020-10-10 20:15:17 2020-10-09  ...               16   
1  2020-10-09 20:14:10  2020-10-10 20:15:17 2020-10-09  ...               80   
2  2020-10-09 20:14:10  2020-10-10 20:15:17 2020-10-09  ...              234   
3  2020-10-09 20:14:10  2020-10-10 20:15:17 2020-10-09  ...              478   
4  2020-10-09 20:14:10  2020-10-10 20:15:17 2020-10-09  ...              234   

  frgnd_proc_name  frgnd_proc_dt           frgnd_proc_ts  \
0             NaN     2020-10-09 2020-10-09 22:00:25.561   
1      chrome.exe     2020-10-09 2020-10-09 22:28:43.737   
2      chrome.exe     2020-10-09 2020-10-09 22:50:41.240   
3             NaN     2020-10-09 2020-10-09 20:14:48.094   
4      chrome.exe     2020-10-09 2020-10-09 22:50:41.240   

  frgnd_proc_duration_ms frgnd_captioned frgnd_window_mode  \
0                      0             NaN               NaN   
1                 802973              c1                 W   
2                2343497              c1                 W   
3                      0             NaN               NaN   
4                2343497              c1                 W   

   frgnd_non_responsive               proc_name proc_cpu_utilization  
0                     f             DSATray.exe                 0.12  
1                     f  GameScannerService.exe                 0.39  
2                     f  GameScannerService.exe                 0.85  
3                     f     MoUsoCoreWorker.exe                 1.12  
4                     f             MsMpEng.exe                 0.58  

[5 rows x 23 columns]

In [94]:
with ProgressBar():
    frgnd_backgrnd_guid_value_count = frgnd_backgrnd_clean_stage2['guid'].value_counts().compute()

[########################################] | 100% Completed | 44.5s


In [98]:
frgnd_backgrnd_guid_value_count

0af707e7d51e4fb6812a596de9fef679    34419
9c4e19608ece40f5b1db5b37d0091a2f    31844
007729cb039e4e1aaa6c4a86c8c436e5    30205
8189859393774a879a54ce8aaeebc181    28970
c69315235fa14225b0cfbc6dbb999e4e    27946
                                    ...  
db970d2a18d94ceb9f1ef1da8b178577       10
dcb2d536665a46daaa3d2151a546bf4d       10
de941903d7474f33bf9243b40cad4766       10
c172bd39a6884990a227743bd57e54de       10
3c07024d3df040e29d72d2f25a86e96e        9
Name: guid, Length: 2083, dtype: int64

In [102]:
frgnd_backgrnd_guid_value_count.describe()

count     2083.000000
mean      1985.278445
std       3733.666166
min          9.000000
25%        130.000000
50%        519.000000
75%       2112.000000
max      34419.000000
Name: guid, dtype: float64

#### Finding2: There are 2083 unique identitfier system in frgnd_backgrnd. Half of the data are less than 519 data point, 75 percent of the uid have less than 2112 data points.

In [116]:
frgnd_backgrnd_clean_stage2[frgnd_backgrnd_apps_poi].head()

event_dt  event_duration_ms      event_name                event_ts  \
0 2020-10-09             156317  DC_DISPLAY_OFF 2020-10-09 22:00:25.561   
1 2020-10-09            5203164   DC_DISPLAY_ON 2020-10-09 22:03:01.879   
2 2020-10-09            5203164   DC_DISPLAY_ON 2020-10-09 22:03:01.879   
3 2020-10-09            4773068   DC_DISPLAY_ON 2020-10-09 20:14:48.094   
4 2020-10-09            5203164   DC_DISPLAY_ON 2020-10-09 22:03:01.879   

  frgnd_captioned frgnd_non_responsive frgnd_proc_dt  frgnd_proc_duration_ms  \
0             NaN                    f    2020-10-09                       0   
1              c1                    f    2020-10-09                  802973   
2              c1                    f    2020-10-09                 2343497   
3             NaN                    f    2020-10-09                       0   
4              c1                    f    2020-10-09                 2343497   

  frgnd_proc_name           frgnd_proc_ts frgnd_window_mode  num_proc_samples  \
0             NaN 2020-10-09 22:00:25.561               NaN                16   
1      chrome.exe 2020-10-09 22:28:43.737                 W                80   
2      chrome.exe 2020-10-09 22:50:41.240                 W               234   
3             NaN 2020-10-09 20:14:48.094               NaN               478   
4      chrome.exe 2020-10-09 22:50:41.240                 W               234   

   proc_cpu_utilization               proc_name  
0                  0.12             DSATray.exe  
1                  0.39  GameScannerService.exe  
2                  0.85  GameScannerService.exe  
3                  1.12     MoUsoCoreWorker.exe  
4                  0.58             MsMpEng.exe

In [104]:
with ProgressBar():
    frgnd_backgrnd_proc_name_value_count = frgnd_backgrnd_clean_stage2['proc_name'].value_counts().compute()

[########################################] | 100% Completed | 42.9s


In [105]:
frgnd_backgrnd_proc_name_value_count

System                                   337369
dwm.exe                                  329388
svchost.exe                              256764
chrome.exe                               219418
explorer.exe                             184487
                                          ...  
install_182053056241.exe                      1
GrabCADPrintInstaller.exe                     1
GrabCADPrintFDMBuildServiceWorker.exe         1
GrabCADPrint.exe                              1
install_25404320341.exe                       1
Name: proc_name, Length: 7586, dtype: int64

In [106]:
frgnd_backgrnd_proc_name_value_count.describe()

count      7586.000000
mean        545.127208
std        8254.199151
min           1.000000
25%           1.000000
50%           5.000000
75%          30.000000
max      337369.000000
Name: proc_name, dtype: float64

#### Finding3: There are 7586 unique process. 75 percent of the process occur 30 times or less in our dataset. The top 5 are System, dwm.exe, svchost.exe, chrome.exe and explorer.exe

##### dwm.exe(Desktop windwo Manager) 
Desktop Window Manager (dwm.exe) is a compositing window manager that renders all those pretty effects in Windows: transparent windows, live taskbar thumbnails, Flip3D, and even high resolution monitor support.



In [117]:
frgnd_backgrnd_clean_stage2["num_proc_samples"].describe().compute()

count    4.135335e+06
mean     5.077163e+01
std      1.337637e+02
min      1.000000e+00
25%      1.100000e+01
50%      2.100000e+01
75%      5.300000e+01
max      7.053000e+03
Name: num_proc_samples, dtype: float64

#### Question1: What is "num_proc_samples" column?

#### frgnd_non_responsive column

In [118]:
with ProgressBar():
    frgnd_non_responsive_value_count = frgnd_backgrnd_clean_stage2['frgnd_non_responsive'].value_counts().compute()

[########################################] | 100% Completed | 40.2s


In [119]:
frgnd_non_responsive_value_count

f    4130078
t       5257
Name: frgnd_non_responsive, dtype: int64

In [123]:
frgnd_backgrnd_clean_stage2.loc[frgnd_backgrnd_clean_stage2["frgnd_non_responsive"] == "t"]["proc_name"].value_counts().compute()

System                394
dwm.exe               329
svchost.exe           301
Memory Compression    248
explorer.exe          217
                     ... 
MNEService.exe          1
avgnt.exe               1
basilisk.exe            1
bdservicehost.exe       1
SavService.exe          1
Name: proc_name, Length: 561, dtype: int64

#### Question2: What is event_duration_ms and frgnd_proc_duration_ms? What are the difference?

### event_name

In [121]:
frgnd_backgrnd_clean_stage2["event_name"].value_counts().compute()

DC_DISPLAY_ON     3763221
DC_DISPLAY_OFF     372114
Name: event_name, dtype: int64

In [47]:
#devuse1_clean_stage2.head()

load_ts         batch_id  \
0  2019-11-14 04:07:26  20191113-200224   
1  2019-11-14 04:07:26  20191113-200224   
2  2019-11-14 04:07:26  20191113-200224   
3  2019-11-14 04:07:26  20191113-200224   
4  2019-11-14 04:07:26  20191113-200224   

                                           audit_zip  \
0  2019111402-i-0728cf8ab79db9e58-fUgBvxtUfADzdPI...   
1  2019111402-i-0728cf8ab79db9e58-fUgBvxtUfADzdPI...   
2  2019111402-i-0728cf8ab79db9e58-fUgBvxtUfADzdPI...   
3  2019111402-i-0728cf8ab79db9e58-fUgBvxtUfADzdPI...   
4  2019111402-i-0728cf8ab79db9e58-fUgBvxtUfADzdPI...   

                audit_internal_path                              guid  \
0  V8DEVUSE4KNOWN_20191113183250.V8  0000e65f586a467ca242d6932782e01f   
1  V8DEVUSE4KNOWN_20191113183250.V8  0000e65f586a467ca242d6932782e01f   
2  V8DEVUSE4KNOWN_20191113183250.V8  0000e65f586a467ca242d6932782e01f   
3  V8DEVUSE4KNOWN_20191113183250.V8  0000e65f586a467ca242d6932782e01f   
4  V8DEVUSE4KNOWN_20191113183250.V8  0000e65f586a467ca242d6932782e01f   

    interval_start_utc     interval_end_utc interval_local_start  \
0  2019-11-11 00:21:39  2019-11-14 02:33:37  2019-11-10 16:21:39   
1  2019-11-11 00:21:39  2019-11-14 02:33:37  2019-11-10 16:21:39   
2  2019-11-11 00:21:39  2019-11-14 02:33:37  2019-11-10 16:21:39   
3  2019-11-11 00:21:39  2019-11-14 02:33:37  2019-11-10 16:21:39   
4  2019-11-11 00:21:39  2019-11-14 02:33:37  2019-11-10 16:21:39   

    interval_local_end                   ts          dt  \
0  2019-11-13 18:33:37  2019-11-10 20:58:51  2019-11-10   
1  2019-11-13 18:33:37  2019-11-11 21:00:33  2019-11-11   
2  2019-11-13 18:33:37  2019-11-11 21:00:33  2019-11-11   
3  2019-11-13 18:33:37  2019-11-11 21:00:33  2019-11-11   
4  2019-11-13 18:33:37  2019-11-11 21:00:33  2019-11-11   

                              device  \
0  GUID_DEVINTERFACE_DISPLAY_ADAPTER   
1         GUID_DEVINTERFACE_KEYBOARD   
2            GUID_DEVINTERFACE_MOUSE   
3                GUID_DEVICE_BATTERY   
4              GUID_DEVINTERFACE_HID   

                                        hw_name  \
0  PCI\VEN_8086&DEV_0F31&SUBSYS_14DD1043&REV_0E   
1                         ACPI\VEN_ATK&DEV_3001   
2                         ACPI\VEN_ETD&DEV_0108   
3                         ACPI\VEN_PNP&DEV_0C0A   
4                          HID\VEN_ATK&DEV_4001   

                                              name  duration  status  
0                             Intel(R) HD Graphics     16571       4  
1       PC/AT Enhanced PS/2 Keyboard (101/102-Key)     82117       4  
2                                    ASUS Touchpad     82117       4  
3  Microsoft ACPI-Compliant Control Method Battery     82117       4  
4            HID-compliant wireless radio controls     82117       4

In [49]:
#device1_guid_value_count

aac4a19cfed947489e7325756559a222    62531
a16fc9dae5e54c388f47d2ba559b98ef    38673
0161e19863ee4d59a4f6ed25420bd3b8    35927
0194bebaef2c458daf29f64271e440db    34160
00f8207451df4a65b7b711efbbd5dae4    34120
                                    ...  
043579b1347f4a159e89c2740a76a75f       59
001a032567ef4f9eb7a9d9757a025ee1       49
b6d98b7fc7f74c568a4d99001bbc41ea       48
7682073d644348269f6ba3e3d63d241d       31
7037a210aebc4bb7ac91294793dfcf74        1
Name: guid, Length: 64843, dtype: int64

In [124]:
#with ProgressBar():
#    device1_guid_value_count = devuse1_clean_stage2['guid'].value_counts().compute()